gist.go

```diff
@@ -0,0 +1,86 @@
+package model
+
+// Typed constants
+type ChannelType string
+
+// CamelCase Constant Names
+const (
+	ChannelTypeOpen    ChannelType = "O"
+	ChannelTypePrivate ChannelType = "P"
+)
```

Comment:
We don't need to expose the type, this would reduce the API clutter.

```suggestion
// Typed constants
type channelType string

// CamelCase Constant Names
const (
	ChannelTypeOpen    channelType = "O"
	ChannelTypePrivate channelType = "P"
)
```

Reply:
We often check the channel type elsewhere, e.g. in the `api4` package. Wouldn't this be harder if we didn't export the type, or would it not matter as long as the constant values are exported?

Reply:
It doesn't matter as long as the constant values are exported. You can still do equal `if `checks, `switch` cases etc.

Reply:
I would advise against not exporting the `ChannelType` here. My reasoning is 
even though the constant themselves will be exported the client of this API will _not_ be able to do something like:

```
func DoSomething(cType ChannelType) error {
	if cType == ChannelTypeOpen {
		// do something
	}
	return nil
}
```

And will be forced to do:
```
func DoSomething(cType string) error {
	if cType == string(ChannelTypeOpen) {
		// do something
	}
	return nil
}
```
Which defeats the whole purpose of having a strong type here.

Reply:
I agree. Not exporting the ChannelType will make the godocs harder to read and will have problems like mentioned above.

Reply:
Looks like I pulled the trigger a bit too quickly on this one. @isacikgoz any thoughts on the latest comments before I revert?

Reply:
@ali-farooq0 may be right, but I'm not sure if you ever use a function like `func DoSomething(cType ChannelType) error`. This is a *property* of an entity, which means you shouldn't pass it to random function and do something with it.

If I had to do such thing, I could do it like this: (which seems more idiomatic to me)
```golang
func DoSomething(c *model.Channel) error {
	switch c.Type {
		// do something
	}
	return nil
}
```

Reply:
As for the docs, @agnivade I don't think the `channelType` is worth to document, `Channel.Type` is. And also there is no obstacles to document `channelType` constants. Anyway I am 2/5 on this, if you are above 2/5 feel free to revert 👍

Reply:
I'm completely 0/5 on this - will leave it to you guys to try and find a consensus :)

Reply:
> This is a property of an entity, which means you shouldn't pass it to random function and do something with it.

Why not? I would argue that it's better to pass in just a property rather than the whole entity to prevent the function from unintentionally mutating the entity and go beyond its scope of concern. However, I do realize that this is just an opinion and both ways are reasonable. 

The big concern here is that we'll be forcing clients to follow a certain pattern where there is no advantage and will cause them grief if they ever decide to create a variable with this type.

@isacikgoz is your concern that we want to prevent clients from creating arbitrary `ChannelType` values? If so then I hope the validation checks that we put in place should take care of that. Even if we don't export this type, we should still validate the `ChannelType` values in our code.

There is a good discussion on [stackoverflow](https://stackoverflow.com/a/37386119/5551762) around this. 

3/5 that we should export this type and the likes thereof.

Reply:
@ali-farooq0 thanks for your clarification. Now I have a better view for your concerns. First of all I consider a `channelType` as itself is not a meaningful entity, it belongs to a `Channel`. It is not a type that you want to use alone, or even create a new one. It's an entity specific property. Clients shouldn't be using `channelType` without a channel. Why they are going to use it?

If you want to use `channelType`, you probably going to need another channel property as well. And we are designing immutable entities, it would be much clear if we reduce the type pool of the functions. I could agree with you if we had mutable entities BTW.

My main concern is adding unnecessary entity to the API. It is already quite big.

Reply:
Ah I see. I agree that `channelType` on its own is not meaningful but since we're exporting the struct through which you can interact with the field _directly_ it feels really odd not to expose its type without any good reason.

Also if exposing too many entities is the only concern in the API package then that feels like a design issue / pkg organization issue and shouldn't affect what we export/not-export. 

Another case where we would need to expose this: Say if a client decides to write a filtering function, that accepts filtering parameters, one of which is filter by type of channel, unless we create that struct and put it inside this very package, it would not be possible for them to accept a type.

```go
type ChannelGetOptions struct {
	ChannelType channelType // unless this struct resides within the same package, this will not be possible.
}
```

Reply:
If there are use cases like this that you are encountering, yeah it might be a reason to export. But whenever you export something you are giving up a bit on your freedom while maintaining the API. 

BTW, technically you can still do the filtering by casting `channelType` constant to `string` or vice versa :)

Reply:
> BTW, technically you can still do the filtering by casting channelType constant to string or vice versa :)

heh true :D, however at that point we've lost type safety which is what I'm arguing for that we should export it precisely because of that reason :)

Reply:
But then we will need to add checks to methods like `channel.SetType(ChannelType)` since someone can simply inject their own `channelType` right? (by doing a `channel.SetType(ChannelType("blabla")`). To make this method safe, you will need to check the type, and add an `error` return to the method.

In the end it is a trade-off, but I would prefer to stay safe on the API side. :)

Reply:
Ahh i think even with restricting the type you won't get type safety.

For example this is perfectly valid.

```
package main

import (
	"fmt"

	"github.com/mattermost/gosample/model"
)

func main() {
	channel := model.Channel{
		Type: "meow",
	}
	fmt.Println("channel:", channel)
}
```

```
package model

type channelType string

const (
	ChannelTypeOpen    channelType = "O"
	ChannelTypePrivate channelType = "P"
)

type Channel struct {
	Type channelType
}
```

```
// output
channel: {meow}
```

or this
```
package main

import (
	"fmt"

	"github.com/mattermost/gosample/model"
)

func main() {
	cType := model.ChannelTypePrivate
	cType = "meow"
	channel := model.Channel{
		Type: cType,
	}
	fmt.Println("channel:", channel)
}
```

These examples compile and run file. Which demonstrates that we will have to  validate types regardless. This is mostly around ease of developer really as types help indicate where to look etc.

Reply:
Yeah, note that the field `Type` of the channel is public, we shouldn't be getting to this point with the new approach. As said before it always comes into trade-offs. However, my suggestion is to make it private in the beginning, if there is a need it wouldn't be a breaking change to make it public. If it is a actual need for today, we can make it from the beginning. (but it shouldn't be IMHO) :)

Reply:
Even if we have getters/setters, we'll still have this validation problem regardless if `channelType` is public or private.

```
package model

type channelType string

type Channel struct {
	cType channelType // private field
}

func (c *Channel) SetType(t channelType) {
	c.cType = t
}

func (c *Channel) Type() channelType {
	return c.cType
}
```


```
func main() {
	// cType := model.ChannelTypePrivate
	// cType = "meow"
	channel := model.Channel{}
	
	channel.SetType("myType") // Can still set the type to any arbitrary string
	fmt.Println("channel:", channel) // outputs channel: {myType}
	fmt.Printf("channel: %T - %t", channel.Type(), channel.Type()) // channel: model.channelType - %!t(model.channelType=myType)
}
```
So don't think we can be safe either way :(

Regarding your concern about adding `error` to return in these type of functions, we can decide to panic if the type isn't correct (a bit harsh but still a viable option)

My point is that there is no real advantage of keeping `channelType` private, it'll only cause confusion in some and annoyance in other cases. So lets keep it `ChannelType` :D

Reply:
So, yeah if someone is ever gonna *need* to use our model types out of model package we can export it. IMO they *shouldn't* use them.

I'm still 2/5 on that, if you are still above it I'm not going to object. Just consider that we can start with making them private, if we see a good reason to make them public we can.

PS: We should never use `panic` in an API :)

Reply:
I wouldn't say never use `panic` :P, but rather it depends on the use case. Also I never said we _should_ use `panic`, it was only an option :)

There are plenty of places in std go library where they use panic. [JSON](https://github.com/golang/go/blob/master/src/encoding/json/encode.go#L307) is one example. But anyway, I do agree that our case isn't extreme that it warrants a panic so I agree with you on that one.

heh, i'm definitely 3/5 on this one (i.e. we should keep the type as public `ChanneType` and not `channelType`). I'm open to having someone else chime in to break the tie :D

Reply:
@agnivade, I'm curious if you have context about what's considered idiomatic Go in this context? I note that `golint` would complain about something similar in the context of "constructors", e.g.:
```
main.go:7:22: exported func NewHiddenType returns unexported type main.hiddenType, which can be annoying to use
```

but it doesn't seem to  complain about the same in the context of public properties.

Reply:
There is no strictly defined convention on this. But usually all code that I have seen exports the type too. I searched the standard library, and could find just one example where the type isn't exported, but the constants were just used inside a single function.

At the end of the day, it's all about the API surface we want to expose, how do we want our docs to look and API usage.

If we dont export it, then we must try to ensure that nobody ever passes just the type of the channel, and always they need to pass the entire channel. They _could_ pass the type, but then there's no static type checking. Not saying it's good or bad. On one hand, the type is a private field, and given we are using immutables, we should usually pass the entire channel. But it's hard to say 1-2 years down the line how will the codebase look.

And there's also the part of the godocs. If we export it, go/doc groups the constants under the type which makes things a bit nicer:

![original](https://user-images.githubusercontent.com/1774000/80729882-7b304300-8b26-11ea-9215-6f171b185815.png)

vs 

![new](https://user-images.githubusercontent.com/1774000/80729894-7e2b3380-8b26-11ea-883c-9308612e3c35.png)

The tradeoff is how much are we gaining by not exporting the type, and is it worth the gain. IMO we should choose the option that makes the codebase readable and easy to maintain in the long-run. But yes, generally speaking, the type is always exported. That's my thoughts from an idiomatic Go perspective. I will leave it to you to come to a decision. :)